src/compiler/core.ts

```diff
@@ -242,8 +259,14 @@ namespace ts {
             const count = array.length;
             if (count > 0) {
                 let pos = 0;
-                let result = arguments.length <= 2 ? array[pos] : initial;
-                pos++;
+                let result: T | U;

```

Comment:
I would cast initial upfront like so:

```ts
let result: <U>initial;
if (arguments.length <= 2) {
  ...
```

This gets rid of the casts elsewhere in the function too